File: frontend/src/types/vencimiento.ts
export type TipoVencimiento =
  | "IVA" | "IIBB" | "GANANCIAS" | "BIENES_PERSONALES" | "CARGAS_SOCIALES" | "SUELDOS"
  | "CONTADOR" | "TARJETA" | "SUSCRIPCION" | "PRESTAMO" | "PLAN_DE_PAGO" | "PAGO_AUTOMATICO" | "OTRO"

export type TipoRecurrencia = "UNICA" | "MENSUAL" | "ANUAL" | "PERSONALIZADA"

/** VENCIDO nunca se persiste: lo calcula el backend en lectura (PENDIENTE + fecha pasada). */
export type EstadoVencimiento = "PENDIENTE" | "VENCIDO" | "PAGADO" | "REPROGRAMADO" | "CANCELADO"

/** Subconjunto que sí acepta el filtro del backend (columna real, sin VENCIDO). */
export type EstadoVencimientoFiltro = "PENDIENTE" | "PAGADO" | "REPROGRAMADO" | "CANCELADO"

export const TIPOS_VENCIMIENTO: TipoVencimiento[] = [
  "IVA", "IIBB", "GANANCIAS", "BIENES_PERSONALES", "CARGAS_SOCIALES", "SUELDOS",
  "CONTADOR", "TARJETA", "SUSCRIPCION", "PRESTAMO", "PLAN_DE_PAGO", "PAGO_AUTOMATICO", "OTRO",
]

export const RECURRENCIAS: TipoRecurrencia[] = ["UNICA", "MENSUAL", "ANUAL", "PERSONALIZADA"]

export const ESTADOS_VENCIMIENTO: EstadoVencimiento[] = ["PENDIENTE", "VENCIDO", "PAGADO", "REPROGRAMADO", "CANCELADO"]

export const ESTADOS_VENCIMIENTO_FILTRO: EstadoVencimientoFiltro[] = ["PENDIENTE", "PAGADO", "REPROGRAMADO", "CANCELADO"]

export type Vencimiento = {
  id: number
  descripcion: string
  tipo: TipoVencimiento
  fecha: string
  importeEstimado: number | null
  monedaId: number
  monedaCodigo: string
  recurrencia: TipoRecurrencia
  intervaloDiasPersonalizado: number | null
  estado: EstadoVencimiento
  cuentaContableId: number | null
  cuentaContableCodigo: string | null
  proveedorId: number | null
  proveedorNombre: string | null
  tarjetaCreditoId: number | null
  tarjetaCreditoEntidad: string | null
  proyectoId: number | null
  proyectoNombre: string | null
  conceptoRecurrenteId: number | null
  conceptoRecurrenteNombre: string | null
  asientoVinculadoId: number | null
  asientoVinculadoNumero: number | null
  origenGeneracion: string
  observaciones: string | null
  motivoCancelacion: string | null
}

export type VencimientoCrearInput = {
  descripcion: string
  tipo: TipoVencimiento
  fecha: string
  importeEstimado?: number
  monedaId: number
  recurrencia: TipoRecurrencia
  intervaloDiasPersonalizado?: number
  cuentaContableId?: number
  proveedorId?: number
  tarjetaCreditoId?: number
  proyectoId?: number
  conceptoRecurrenteId?: number
  observaciones?: string
}

export type VencimientoEditarInput = VencimientoCrearInput

export type VencimientoBusquedaFiltros = {
  tipo?: TipoVencimiento
  estado?: EstadoVencimientoFiltro
  fechaDesde?: string
  fechaHasta?: string
  proyectoId?: number
  proveedorId?: number
  tarjetaId?: number
}

export type GenerarAutomaticosResponse = {
  generados: { origen: string; cantidad: number }[]
  total: number
}

File: backend/src/main/java/com/montanaritech/contable/vencimientos/Vencimiento.java
package com.montanaritech.contable.vencimientos;

import com.montanaritech.contable.common.tenant.EntidadNegocio;
import com.montanaritech.contable.contabilidad.asiento.Asiento;
import com.montanaritech.contable.contabilidad.cuentacontable.CuentaContable;
import com.montanaritech.contable.impuestos.atribucion.TipoLiquidacion;
import com.montanaritech.contable.maestros.concepto.Concepto;
import com.montanaritech.contable.maestros.moneda.Moneda;
import com.montanaritech.contable.maestros.proveedor.Proveedor;
import com.montanaritech.contable.maestros.proyecto.Proyecto;
import com.montanaritech.contable.maestros.tarjetacredito.TarjetaCredito;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDate;
import lombok.Getter;
import lombok.Setter;

/**
 * Obligación de pago futura (F8.1): impuestos, tarjeta, sueldos, suscripciones,
 * etc. Se genera a mano o automáticamente (ver {@code VencimientoService.
 * generarAutomaticos}, disparado on-demand desde el frontend — el proyecto no
 * tiene infraestructura de scheduling y este paso no la introduce).
 *
 * <p>{@code estado} solo guarda PENDIENTE/PAGADO/REPROGRAMADO/CANCELADO;
 * VENCIDO se deriva en lectura comparando {@code fecha} contra hoy cuando
 * {@code estado == PENDIENTE}, mismo criterio que {@code EstadoVencimiento}
 * de F4.5 (F3.6/F4.5 nunca persisten ese estado, siempre lo recalculan).
 *
 * <p>{@code liquidacionTipo}/{@code liquidacionId} es una referencia
 * polimórfica sin FK real (mismo patrón que {@code AtribucionImpuesto} de
 * F6.3): sirve para IVA y para IIBB sin tocar esas entidades.
 */
@Entity
@Table(name = "vencimiento")
@Getter
@Setter
public class Vencimiento extends EntidadNegocio {

    @Column(nullable = false, length = 200)
    private String descripcion;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private TipoVencimiento tipo;

    @Column(nullable = false)
    private LocalDate fecha;

    @Column(name = "importe_estimado", precision = 18, scale = 2)
    private BigDecimal importeEstimado;

    @ManyToOne(optional = false)
    @JoinColumn(name = "moneda_id", foreignKey = @ForeignKey(name = "fk_vencimiento_moneda"))
    private Moneda moneda;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private TipoRecurrencia recurrencia = TipoRecurrencia.UNICA;

    @Column(name = "intervalo_dias_personalizado")
    private Integer intervaloDiasPersonalizado;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private EstadoVencimientoObligacion estado = EstadoVencimientoObligacion.PENDIENTE;

    @ManyToOne
    @JoinColumn(name = "cuenta_contable_id", foreignKey = @ForeignKey(name = "fk_vencimiento_cuenta_contable"))
    private CuentaContable cuentaContable;

    @ManyToOne
    @JoinColumn(name = "proveedor_id", foreignKey = @ForeignKey(name = "fk_vencimiento_proveedor"))
    private Proveedor proveedor;

    @Enumerated(EnumType.STRING)
    @Column(name = "liquidacion_tipo", length = 10)
    private TipoLiquidacion liquidacionTipo;

    @Column(name = "liquidacion_id")
    private Long liquidacionId;

    @ManyToOne
    @JoinColumn(name = "tarjeta_credito_id", foreignKey = @ForeignKey(name = "fk_vencimiento_tarjeta_credito"))
    private TarjetaCredito tarjetaCredito;

    @ManyToOne
    @JoinColumn(name = "proyecto_id", foreignKey = @ForeignKey(name = "fk_vencimiento_proyecto"))
    private Proyecto proyecto;

    @ManyToOne
    @JoinColumn(name = "concepto_recurrente_id", foreignKey = @ForeignKey(name = "fk_vencimiento_concepto_recurrente"))
    private Concepto conceptoRecurrente;

    /** Al marcar pagado, vínculo opcional al asiento real ya generado por el Cobro/Pago/PagoTarjeta correspondiente. */
    @ManyToOne
    @JoinColumn(name = "asiento_vinculado_id", foreignKey = @ForeignKey(name = "fk_vencimiento_asiento_vinculado"))
    private Asiento asientoVinculado;

    @Enumerated(EnumType.STRING)
    @Column(name = "origen_generacion", nullable = false, length = 30)
    private OrigenGeneracionVencimiento origenGeneracion = OrigenGeneracionVencimiento.MANUAL;

    @Column(name = "origen_generacion_ref_id")
    private Long origenGeneracionRefId;

    @Column(length = 500)
    private String observaciones;

    @Column(name = "motivo_cancelacion", length = 500)
    private String motivoCancelacion;
}
